diff --git a/src/displayapp/DisplayApp.cpp b/src/displayapp/DisplayApp.cpp
index b1594f19..b966883b 100644
--- a/src/displayapp/DisplayApp.cpp
+++ b/src/displayapp/DisplayApp.cpp
@@ -476,7 +476,16 @@ void DisplayApp::Refresh() {
         motorController.RunForDuration(35);
         break;
       case Messages::OnChargingEvent:
-        motorController.RunForDuration(15);
+        if (batteryController.IsCharging() && currentApp == Apps::Clock) {
+            // Open the battery app if on the clock screen
+            LoadNewScreen(Apps::BatteryInfo, DisplayApp::FullRefreshDirections::None);
+        } else if (!batteryController.IsCharging() && currentApp == Apps::BatteryInfo) {
+            // Close the battery app after being unplugged
+            LoadNewScreen(Apps::Clock, DisplayApp::FullRefreshDirections::None);
+        } else {
+            // Vibrate normally otherwise as to not close any open app
+            motorController.RunForDuration(15);
+        }
         break;
     }
   }
